web/src/main/webapp/ui/menu.ts

```diff
@@ -40,157 +41,279 @@ export interface MenuItem extends BaseMenuItem {
     readonly action: () => void;
 }
 
-abstract class BaseMenu<MI extends BaseMenuItem> implements IHtmlElement {
-    public items: MI[];
-    /**
-     * Actual html representations corresponding to the submenu items.
-     * They are stored in the same order as the items.
-     */
-    public outer: HTMLTableElement;
-    public tableBody: HTMLTableSectionElement;
-    public cells: HTMLTableDataCellElement[];
-    public selectedIndex: number;  // -1 if no item is selected
-
-    protected constructor() {
-        this.items = [];
-        this.cells = [];
-        this.selectedIndex = -1;
-        this.outer = document.createElement("table");
-        this.outer.classList.add("menu", "hidden");
-        this.tableBody = this.outer.createTBody();
-        this.outer.onkeydown = (e) => this.keyAction(e);
-    }
-
-    protected addItems(mis: MI[]): void {
-        if (mis != null) {
-            for (const mi of mis)
-                this.addItem(mi, true);
-        }
-    }
-
-    public abstract setAction(mi: MI, enabled: boolean): void;
-
-    public keyAction(e: KeyboardEvent): void {
-        if (e.code === "ArrowDown" && this.selectedIndex < this.cells.length - 1) {
-            this.select(this.selectedIndex + 1);
-        } else if (e.code === "ArrowUp"  && this.selectedIndex > 0) {
-            this.select(this.selectedIndex - 1);
-        } else if (e.code === "Enter" && this.selectedIndex >= 0) {
-            // emulate a mouse click on this cell
-            this.cells[this.selectedIndex].click();
-        } else if (e.code === "Escape") {
-            this.hide();
-        }
-    }
+interface subMenuCells {
+    cells: HTMLTableDataCellElement[];
+}
 
-    /**
-     * Find the position of an item based on its text.
-     * @param text   Text string in the item.
-     */
-    public find(text: string): number {
-        for (let i = 0; i < this.items.length; i++)
-            if (this.items[i].text === text)
-                return i;
-        return -1;
+function emptySubMenu(): subMenuCells {
+    return {
+        cells: []
     }
+}
 
-    public getHTMLRepresentation(): HTMLElement {
-        return this.outer;
+abstract class BaseMenu<MI extends BaseMenuItem> implements IHtmlElement {
```

Comment:
please use the same formatting as everywhere else: four spaces.
Reformatting this makes is very difficult to compare the code.